### ISSUE TITLE
bug: Matern and euclidean_distance

### ISSUE THREAD

**Issue by ShuSheng3927** [JaxGaussianProcesses/GPJax#508]
Henry (@henrymoss ) and I spotted a strange implementation detail about the stationary kernel. 

In gpjax/kernels/stationary/utils.py, the helper function 

def euclidean_distance(x: Float[Array, " D"], y: Float[Array, " D"]) -> ScalarFloat:
    r"""Compute the euclidean distance between a pair of inputs.

    Args:
        x (Float[Array, " D"]): First input.
        y (Float[Array, " D"]): Second input.

    Returns
    -------
        ScalarFloat: The euclidean distance between the inputs.
    """
    return jnp.sqrt(jnp.maximum(squared_distance(x, y), 1e-36))

uses "jnp.maximum(squared_distance(x, y), 1e-36)". When we try to use jax.grad (or jax.hessian) to evaluate the derivate at points where x=y, this additional (numerical stability) hack yields wrong gradients. 

What might be the motivation for this implementation in the beginning? 

Also a typo in https://docs.jaxgaussianprocesses.com/api/kernels/stationary/matern52/ and https://docs.jaxgaussianprocesses.com/api/kernels/stationary/matern32/: the equation has an unnecessary "exp" in the middle term.

**Comment by thomaspinder** [comment]
I'm not sure I agree that this is a bug. As far as I am aware, the clipping of the euclidean distance is required in a Matérn kernel $k$ for stable gradients. The values of $x$ and $y$ depend on the lengthscale $\ell$. If I wish to compute a derivative of the mll w.r.t. $\ell$, then jax will unroll the function into a sequence of elementary functions in a graph and then compute the chain rule. One of the functions that I will be required to compute a partial derivative of and then evaluate ends up having |x-y| as the denominator. Without the clipping, this function would be undefined, and your gradient would not evaluate. 

I may be wrong on this - its been some time (over 2 years according to Git blame) since I thought about this, but this is the reason as I recall. If you disagree, or feel that there is a better soln. to this issue, then please do followup!

Thanks for the flagging the docs issue - I will address separately.

**Comment by ShuSheng3927** [comment]
Thanks for your quick reply! 

I do agree that if one removes the clipping and does jnp.sqrt(euclidean_distance(x,y)) directly, it would indeed give NaN when x == y. However, at least in the 1d case, doing jnp.abs(x-y) would bypass this issue. 

Further, the exact corner case where we ran into an issue is when we compute partial_x partial_y k(x,y) using jax.hessian. See below for details. 

`def matern_kernel1(x,y):
    l = 1
    sig2 = 1
    tau = jnp.abs((x - y) / l)

    return sig2 * (1 + jnp.sqrt(3)*tau) * jnp.exp(- jnp.sqrt(3) * tau) 

def matern_kernel2(x,y):
    l = 1
    sig2 = 1
    tau = jnp.sqrt(jnp.maximum(jnp.sum((x-y)**2) / l, 1e-36))

    return sig2 * (1 + jnp.sqrt(3)*tau) * jnp.exp(- jnp.sqrt(3) * tau) 

kernel_hessian1 = jax.hessian(matern_kernel1, argnums=(0,1))
kernel_hessian2 = jax.hessian(matern_kernel2, argnums=(0,1))

print(kernel_hessian1(jnp.array([0.]), jnp.array([0.])))
#correct
#((Array([[[-3.]]], dtype=float64), Array([[[3.]]], dtype=float64)), (Array([[[3.]]], dtype=float64), Array([[[-3.]]], dtype=float64))) 

print(kernel_hessian2(jnp.array([0.]), jnp.array([0.])))
#incorrect
#((Array([[0.]], dtype=float64), Array([[0.]], dtype=float64)), (Array([[-0.]], dtype=float64), Array([[-0.]], dtype=float64)))`


I am not immediately sure how one would fix this issue for higher dimensional x and y. But I think there would be a similar solution to jnp.abs.

**Comment by thomaspinder** [comment]
Yes, I think the equivalency with `jnp.abs` only holds when x and y are 1-dimensional. 

Whilst I'm sympathetic to the issue that you face here, I think supporting two ways for computing euclidean distances where the approach is conditional upon the inputs' dimensionality would unnecessarily complicate the API. If you are able to identify an approach that works in all dimensions, then I'd gladly accept a pull request for such a function. However, the best short-term solution for you may be to subclass the Matern kernel class here and redefine `__call__` using your own distance fn. therein. 

As an aside, I checked GPyTorch and GPFlow, and it appears that these two packages computed clipped Euclidean distances.

**Comment by henrymoss** [comment]
yeh you are right Tom. We are just using this kernel for time components, so can still with the abs hack. 

Our use case --- needing to back propogate through a kernel that is evalauted at very very close points --- is perhaps a little niche anyway ?
